README.rst
maxtweety README


CHANGES.rst
Changelog
=========

4.0.5 (unreleased)
------------------

- Nothing changed yet.


4.0.4 (2022-04-29)
------------------

* setuptools and ipdb [Pilar Marinas]

4.0.3 (2014-12-01)
------------------

* Updat to new instances.ini format [Carles Bruguera]

4.0.2 (2014-11-25)
------------------

* Provide custom client_properties to pika [Carles Bruguera]

4.0.1 (2014-06-10)
------------------

* Don't listen to empty hashtags [Carles Bruguera]

4.0.0 (2014-04-15)
------------------

* Adapt maxtweety to new rabbitmq architecture [Carles Bruguera]
* Depend on maxcarrot for messages [Carles Bruguera]

1.1.3 (2013-11-07)
------------------

* Auto-restart process on tweepy crash [Carles Bruguera]

1.1.2 (2013-10-17)
------------------

* No limit in max response lists [Carles Bruguera]

1.1.1 (2013-10-03)
------------------

 * Re-enable stream listener [Carles Bruguera]


1.1 (2013-10-03)
----------------
 * Parametrize the delay restart time [Carles Bruguera]
 * Listen to hastags [Carles Bruguera]
 * Delete unused tasks.py Change to use separated ini files Use the correct log files [Carles Bruguera]
 * Avoid too frequent restarts [Carles Bruguera]
 * Change rabbitmq connection parameters method [Carles Bruguera]

1.0 (2013-08-13)
----------------

-  Initial version


maxtweety/tests.py
import unittest

from pyramid import testing


class ViewTests(unittest.TestCase):
    def setUp(self):
        self.config = testing.setUp()

    def tearDown(self):
        testing.tearDown()

    def test_my_view(self):
        from .views import my_view
        request = testing.DummyRequest()
        info = my_view(request)
        self.assertEqual(info['project'], 'maxtweety')


maxtweety/tasks.py
import os
from maxrules.twitter import twitter_generator_name, debug_hashtag, max_server_url
import requests

from maxrules.config import mongodb_url, mongodb_db_name, mongodb_cluster, mongodb_hosts, mongodb_replica_set
from max.MADMax import MADMaxCollection
from max.rest.utils import canWriteInContexts
from max.rest.utils import findHashtags
import json
import logging


# logging_file = '/var/pyramid/maxserver/var/log/twitter-processor.log'

# # Temporary patch to avoid tests failure
# if not os.path.exists(logging_file):  # pragma: no cover
#     logging_file = '/tmp/twitter-processor.log'
logger = logging.getLogger(__name__)
# fh = logging.FileHandler(logging_file, encoding="utf-8")
# formatter = logging.Formatter('%(asctime)s %(message)s')
# logger.setLevel(logging.DEBUG)
# fh.setFormatter(formatter)
# logger.addHandler(fh)


def processTweet(self, twitter_username, content, tweetID='---'):
    """ Process inbound tweet
    """
    # Setup logging
    logger.info(u"(INFO) Processing tweet %s from %s with content: %s" % (str(tweetID), twitter_username, content))

    # DB connection
    if not mongodb_cluster:
        conn = pymongo.MongoClient(mongodb_url)
    else:
        conn = pymongo.MongoReplicaSetClient(mongodb_hosts, mongodb_replica_set)

    db = conn[mongodb_db_name]
    users = MADMaxCollection(db.users)
    contexts = MADMaxCollection(db.contexts)

    contexts_with_twitter_username = contexts.search({"twitterUsernameId": {"$exists": True}})
    readable_follow_list = [users_to_follow.get('twitterUsername', '').lower() for users_to_follow in contexts_with_twitter_username]
    twitter_username = twitter_username.lower()
    # If we have a tweet from a followed user
    if twitter_username in readable_follow_list:
        # Find the context
        maxcontext = contexts.search({"twitterUsername": twitter_username})

        # Watch for the case when two or more context share twitterUsername
        for context in maxcontext:
            url_hash = context.get("hash")
            context_url = context.get("url")

            # Construct the payload with the activity information
            newactivity = {
                "object": {
                    "objectType": "note",
                    "content": content
                },
                "contexts": [{'url': context_url,
                              'objectType': 'context'
                              }
                             ],
                "generator": twitter_generator_name
            }

            re = requests.post('%s/contexts/%s/activities' % (max_server_url, url_hash), json.dumps(newactivity), auth=('admin', 'admin'), verify=False)
            if re.status_code == 201:
                # 200: Successful tweet from context
                logger.info(u"(201) Successfully posted tweet %s from %s as context %s" % (str(tweetID), twitter_username, context_url))
                return u"(201) %s tweet from context %s" % (twitter_username, context_url)
            else:  # pragma: no cover
                # 500: Error accessing the MAX API
                logger.info(u"(500) Error accessing the MAX API at %s" % max_server_url)
                return u"(500) MAX API error"

        return u"Should not see me..."  # pragma: no cover

    # If we have a tweet from a tracked hashtag
    # Parse text and determine the second or nth hashtag
    possible_hastags = findHashtags(content)
    # Normalize possible_hastags
    #possible_hastags = [hashtag.lower() for hashtag in possible_hastags]
    # Prepare query with normalized possible_hastags

    query = [dict(twitterHashtag=hashtag.lower()) for hashtag in possible_hastags]

    if debug_hashtag in possible_hastags:
        logger.info(u"Debug hashtag detected! %s" % content)
        return u"Debug hashtag detected! %s" % content

    # Check if twitter_username is a registered for a valid MAX username
    # if not, discard it
    maxuser = users.search({"twitterUsername": twitter_username})
    if maxuser:
        maxuser = maxuser[0]
    else:
        logger.info(u"(404) Discarding tweet %s from %s : There's no MAX user with that twitter username." % (str(tweetID), twitter_username))
        return u"(404) %s: No such MAX user." % twitter_username

    # Check if hashtag is registered for a valid MAX context
    # if not, discard it

    successful_tweets = 0
    maxcontext = contexts.search({"$or": query})
    if maxcontext:
        for context in maxcontext:
            # Check if MAX username has permission to post to the MAX context
            # if not, discard it
            try:
                can_write = canWriteInContexts(maxuser, [context, ])
            except:
                can_write = False
                logger.info(u"(401) %s can't write to %s" % (maxuser.username, context['url']))

            if can_write:
                # Construct the payload with the activity information
                newactivity = {
                    "object": {
                        "objectType": "note",
                        "content": content
                    },
                    "contexts": [{'url': context['url'],
                                  'objectType': 'context'
                                  }
                                 ],
                    "generator": twitter_generator_name
                }

                # Use the restricted REST endpoint for create a new activity in the specified
                # MAX context in name of the specified MAX username
                re = requests.post('%s/people/%s/activities' % (max_server_url, maxuser.username), json.dumps(newactivity), auth=('admin', 'admin'), verify=False)
                if re.status_code == 201:
                    logger.info(u"(201) Successfully posted tweet %s from @%s as %s on context %s" % (str(tweetID), twitter_username, maxuser['username'], context.url))
                    successful_tweets += 1
                    #return u"Success tweet from user %s in context %s" % (maxuser, context.url)
                else:
                    logger.info(u"(500) Error accessing the MAX API at %s" % max_server_url)  # pragma: no cover
                    #return u"Error accessing the MAX API at %s" % max_server_url
        error_message = len(maxcontext) != successful_tweets and " We weren't able to send the tweet to all contexts. See above for more information." or ""
        logger.info(u"Processed tweet %s to %d of %d possible contexts.%s" % (str(tweetID), successful_tweets, len(maxcontext), error_message))
        if error_message:
            return u"(401) Some posts not sent"
        else:
            return u"(200) All posts sent"
    else:
        logger.info(u"(404) Discarding tweet %s from %s: There are no MAX context with any of those hashtags" % (str(tweetID), twitter_username))
        return u"(404) %s: Not such MAX context %s" % (maxuser.username, maxcontext)

    return u"Should not see mee"  # pragma: no cover


maxtweety/utils.py
from logging.config import fileConfig
import ConfigParser
import os


def _getpathsec(config_uri, name):
    if '#' in config_uri:
        path, section = config_uri.split('#', 1)
    else:
        path, section = config_uri, 'main'
    if name:
        section = name
    return path, section


def setup_logging(config_uri):
    """
    Set up logging via the logging module's fileConfig function with the
    filename specified via ``config_uri`` (a string in the form
    ``filename#sectionname``).

    ConfigParser defaults are specified for the special ``__file__``
    and ``here`` variables, similar to PasteDeploy config loading.
    """
    path, _ = _getpathsec(config_uri, None)
    parser = ConfigParser.ConfigParser()
    parser.read([path])
    if parser.has_section('loggers'):
        config_file = os.path.abspath(path)

        fc = fileConfig(
            config_file,
            dict(__file__=config_file, here=os.path.dirname(config_file))
        )
        return fc


maxtweety/listener.py
from textwrap import TextWrapper

import json
import logging
import pika
import tweepy

from maxcarrot.message import RabbitMessage

logger = logging.getLogger('twitterlistener')


class StreamWatcherListener(tweepy.StreamListener):  # pragma: no cover

    status_wrapper = TextWrapper(width=60, initial_indent='    ', subsequent_indent='    ')

    def __init__(self, rabbit_server, api=None):
        super(StreamWatcherListener, self).__init__(api)
        self.rabbit_server = rabbit_server

    def on_status(self, status):
        # Initialize Rabbit connection
        self.connection = pika.BlockingConnection(
            pika.URLParameters(self.rabbit_server)
        )
        self.channel = self.connection.channel()

        try:
            logger.info('Got tweet %d from %s via %s with content: %s' % (status.id, status.author.screen_name, status.source, status.text))
            # Insert the data of the new tweet into the Rabbit queue
            twitter_message = dict(author=status.author.screen_name.lower(), message=status.text, stid=status.id)

            message = RabbitMessage()
            message.prepare({
                "source": "tweety",
                "version": 4.0,
            })
            message['data'] = twitter_message
            message['object'] = 'tweet'
            message['action'] = 'add'

            self.channel.basic_publish(
                exchange='twitter',
                routing_key='',
                body=json.dumps(message.packed)
            )
            self.connection.close()
        except:
            # Catch any unicode errors while printing to console
            # and just ignore them to avoid breaking application.
            pass

    def on_error(self, status_code):
        logging.error('An error has occured! Status code = %s' % status_code)
        return True  # keep stream alive

    def on_timeout(self):
        logging.warning('Snoozing Zzzzzz')


maxtweety/__init__.py
from maxtweety.listener import StreamWatcherListener
from maxtweety.utils import setup_logging

import argparse
import ConfigParser
import json
import logging
import multiprocessing
import pika
import requests
import pkg_resources
import sys
import tweepy
import arrow
import time
import datetime

debug_hashtag = u'#debugmaxupcnet'
logger = logging.getLogger('tweety')


# PATCH Pika to provide extended client_properties to Rabbitmq Management plugin

@property
def _client_properties(self):
    """Return the client properties dictionary.

    :rtype: dict

    """
    return {'library': pika.connection.PRODUCT,
            'library-version': pkg_resources.require('pika')[0].version,
            'capabilities': {'authentication_failure_close': True,
                             'basic.nack': True,
                             'connection.blocked': True,
                             'consumer_cancel_notify': True,
                             'publisher_confirms': True},
            'information': 'See http://pika.rtfd.org',
            'platform': 'Python {0.major}.{0.minor}.{0.micro}'.format(sys.version_info),
            'product': 'maxtweety',
            'version': pkg_resources.require('maxtweety')[0].version}

pika.connection.Connection._client_properties = _client_properties


class RestartClock(object):
    def __init__(self, seconds_between_restarts=20):
        self.reset()
        self.delay = seconds_between_restarts

    def reset(self):
        """
            Resets the timer to the current time
        """
        self.last = arrow.now()

    def ready(self):
        """
            Returns True if enough time has elapsed since last reset
        """
        elapsed = arrow.now() - self.last
        elapsed_seconds = elapsed.seconds
        return elapsed_seconds >= self.delay

    def remaining(self):
        """
            Time remaining to be ready again
        """
        elapsed = arrow.now() - self.last
        elapsed_seconds = elapsed.seconds
        time_to_wait = self.delay - elapsed_seconds
        return time_to_wait

    def wait(self):
        time.sleep(self.remaining())

    def in_time(self, restart_request):
        return restart_request is None or restart_request > self.last


def main(argv=sys.argv, quiet=False):  # pragma: no cover
    tweety = MaxTwitterListenerRunner(argv, quiet)
    tweety.spawn_process()

    connection = pika.BlockingConnection(
        pika.URLParameters(tweety.common.get('rabbitmq', 'server'))
    )

    channel = connection.channel()

    logger.info('[*] Waiting for restart signal.')

    try:
        restart_delay = int(tweety.config.get('main', 'minimum_restart_delay'))
    except:
        restart_delay = 300

    clock = RestartClock(seconds_between_restarts=restart_delay)

    def callback(ch, method, properties, body):

        received_restart_msg = "[*] Received restart from MAX Server -->"
        try:
            restart_request_time = arrow.get(datetime.datetime.utcfromtimestamp(float(body)))
        except:
            restart_request_time = None
        ch.basic_ack(delivery_tag=method.delivery_tag)

        if clock.in_time(restart_request_time):
            if not clock.ready():
                remaining = clock.remaining()
                next_restart = arrow.now().replace(seconds=remaining).strftime('%H:%M:%S')
                logger.info("{} Delayed until {}".format(received_restart_msg, next_restart))
                clock.wait()
            tweety.restart()
            logger.info("{} Restarted".format(received_restart_msg))
            clock.reset()
        else:
            logger.info("{} Discarding ancient restart".format(received_restart_msg))

    channel.basic_qos(prefetch_count=1)
    channel.basic_consume(callback,
                          queue='tweety_restart')
    channel.start_consuming()


class MaxTwitterListenerRunner(object):  # pragma: no cover
    process = None

    verbosity = 1  # required
    description = "Max Twitter listener runner."
    usage = "usage: %prog [options]"
    parser = argparse.ArgumentParser(description=description)
    parser.add_argument(
        '-c', '--config',
        dest='configfile',
        type=str,
        help=("Configuration file"))

    def __init__(self, argv, quiet=False):
        self.quiet = quiet
        self.options = self.parser.parse_args()
        self.config = ConfigParser.ConfigParser()
        self.config.read(self.options.configfile)

        if not self.options.configfile:
            logging.error('You must provide a valid configuration .ini file.')
            sys.exit(1)

        setup_logging(self.options.configfile)

        common_config_file = self.config.get('main', 'common')
        cloudapis_config_file = self.config.get('main', 'cloudapis')
        instances_config_file = self.config.get('main', 'instances')

        self.common = ConfigParser.ConfigParser()
        self.common.read(common_config_file)

        self.cloudapis = ConfigParser.ConfigParser()
        self.cloudapis.read(cloudapis_config_file)

        self.instances = ConfigParser.ConfigParser()
        self.instances.read(instances_config_file)

        try:
            self.consumer_key = self.cloudapis.get('twitter', 'consumer_key')
            self.consumer_secret = self.cloudapis.get('twitter', 'consumer_secret')
            self.access_token = self.cloudapis.get('twitter', 'access_token')
            self.access_token_secret = self.cloudapis.get('twitter', 'access_token_secret')

            self.maxservers_settings = [maxserver for maxserver in self.instances.sections()]

        except:
            logging.error('You must provide a valid configuration .ini file.')
            sys.exit(1)

    def spawn_process(self, *args, **kwargs):
        self.process = multiprocessing.Process(target=self.run, args=args, kwargs=kwargs)
        self.process.start()

    def restart(self):
        if self.process.is_alive():
            self.process.terminate()
        self.spawn_process()

    def send_restart(self):
        connection = pika.BlockingConnection(
            pika.URLParameters(self.rabbit_server)
        )
        channel = connection.channel()
        restart_request_time = datetime.datetime.now().strftime('%s.%f')
        channel.basic_publish(
            exchange='',
            routing_key='tweety_restart',
            body=restart_request_time)

    def get_twitter_enabled_contexts(self):
        contexts = {}
        for max_server_name in self.maxservers_settings:
            max_url = self.instances.get(max_server_name, 'server')
            max_restricted_user = self.instances.get(max_server_name, 'restricted_user')
            max_restricted_user_token = self.instances.get(max_server_name, 'restricted_user_token')
            req = requests.get('{}/contexts'.format(max_url), params={"twitter_enabled": True, "limit": 0}, headers=self.oauth2Header(max_restricted_user, max_restricted_user_token))
            if req.status_code == 200:
                context_follow_list = [users_to_follow.get('twitterUsernameId') for users_to_follow in req.json() if users_to_follow.get('twitterUsernameId')]
                context_readable_follow_list = [users_to_follow.get('twitterUsername') for users_to_follow in req.json() if users_to_follow.get('twitterUsername')]
                contexts.setdefault(max_server_name, {})['ids'] = context_follow_list
                contexts[max_server_name]['readable'] = context_readable_follow_list
            else:
                logging.error('Failed to get contexts from "{}" at {}: {}'.format(max_server_name, max_url, req.content))

        self.users_id_to_follow = contexts

    def flatten_users_id_to_follow(self):
        flat_list = []
        for maxserver_name in self.users_id_to_follow.keys():
            id_list = self.users_id_to_follow.get(maxserver_name).get('ids')
            flat_list = flat_list + id_list

        return flat_list

    def get_max_global_hashtags(self):
        self.global_hashtags = []
        for max_server_name in self.maxservers_settings:
            hashtag = self.instances.get(max_server_name, 'hashtag', '').strip('')
            # Don't listen to empty hashtags
            if hashtag:
                self.global_hashtags.append('#{}'.format(hashtag))

    def oauth2Header(self, username, token, scope="widgetcli"):
        return {
            "X-Oauth-Token": token,
            "X-Oauth-Username": username,
            "X-Oauth-Scope": scope}

    def run(self):
        self.get_max_global_hashtags()
        self.get_twitter_enabled_contexts()

        # Prompt for login credentials and setup stream object
        auth = tweepy.OAuthHandler(self.consumer_key, self.consumer_secret)
        auth.set_access_token(self.access_token, self.access_token_secret)

        # auth = tweepy.auth.BasicAuthHandler(self.options.username, self.options.password)
        self.rabbit_server = self.common.get('rabbitmq', 'server')
        stream = tweepy.Stream(auth, StreamWatcherListener(self.rabbit_server), timeout=None)

        # Add the debug hashtag
        self.global_hashtags.append(debug_hashtag)

        logger.info("Listening to this Twitter hashtags: {}".format(', '.join(self.global_hashtags)))
        logger.info("Listening to this Twitter users: \n{}".format(json.dumps(dict([(k, ['{:<15} ({})'.format(*a) for a in zip(v['readable'], v['ids'])]) for k, v in self.users_id_to_follow.items()]), indent=4)))

        try:
            stream.filter(follow=self.flatten_users_id_to_follow(), track=self.global_hashtags)
        except:
            logger.info('Stream listener unexpectedly exited, sending auto-restart signal.')
            self.send_restart()
        return
